### DISCUSSION TITLE
Docker

### DISCUSSION THREAD

**Discussion by sOckhamSter** [flobz/psa_car_controller#102] [Ideas]
Hi @flobz - Here's my first attempt at containerisation of psa_car_controller:

https://github.com/sOckhamSter/docker_psa_car_controller
https://hub.docker.com/r/sockhamster/psa_car_controller/

My suggested run command is:
```
  docker run -ti --name psa_car_controller1 \
    --publish 8180:8180 \
    -v /host_path/config:/config \
    psa_car_controller
```

By default, this will start the container with command line access only, so just exec to bash:

`docker exec -it psa_car_controller1 /bin/bash`

... and then all of the standard setup can be done as per your readme. Once the app_decoder work has been completed, the test.json and charge_config1.json files should be moved to /config:

```
mv /psa_car_controller/test.json /config
mv /psa_car_controller/charge_config1.json /config
```

Then start the server for the first time to input the SMS/PIN:

`python3 server.py -f /config/test.json -c /config/charge_config1.json -p 8180 -l 0.0.0.0 -r`

Note: the IP address needs to be the wildcard 0.0.0.0 because it has to bind to whatever address the container has been given, and the port is 8180 because your default of 5000 conflicts with my Synology NAS...

Once it's all working, ctrl-c out of server.py and stop the container. Then edit the /config/dockerconfig.conf file that should be in the mapped folder and change the line to:

`export SHELL_ONLY="FALSE"`

Restart the container and server.py will automatically run as the main service.

Currently I have this sitting on my GitHub repo, but with your permission I'd like to merge those files into your repo. The main reason for that is because we can then enable auto-build on Docker Hub so every time you update your main branch, the Docker image will re-build automatically. Otherwise right now I would have to manually push an update to Docker and there would end up being a lag / SPOF on me.

**Comment by sOckhamSter** [comment]
Just to add to this, the one thing I can't figure out is where the SMS/PIN authentication response is being cached? If that response can somehow be placed in /config then it will persist between container refreshes. At the moment, every time the container is updated, you have to go through the SMS/PIN faff again. Any suggestions welcome!

**Reply by flobz** [reply]
In otp.bin
If you are in another dir than the app dir, can you check where config file are saved? 
It should be in the workdir.

**Reply by sOckhamSter** [reply]
Amazing, thank you. By changing the WORKDIR in Dockerfile, it can now use /config for everything. I've also had to put in a few literal paths to the Python scripts just to make sure everything runs too, but the updated Dockerfile is now:

```
FROM python:3
RUN cd / && git clone https://github.com/flobz/psa_car_controller.git
WORKDIR /config
RUN pip3 install -r /psa_car_controller/requirements.txt
#RUN mkdir /config
##### These lines to be removed when merged with main psa_car_controller repo
RUN mkdir /psa_car_controller/docker_files
COPY docker_files/init.sh /psa_car_controller/docker_files
COPY docker_files/dockerconfig.conf /psa_car_controller/docker_files
##### End of temporary lines
EXPOSE 8180
CMD /psa_car_controller/docker_files/init.sh
```

It all seems to run quite well for me. Would you be happy for me to put a pull request in? The plan is to put the Dockerfile in the root of your project so as Docker Hub can see it, then create a new folder called docker_files for the init.sh script and the template config file. Then I'd need to somehow link my Docker image to your Github repo for auto builds. I've had a quick look at this and it seems I can't add in any repo that isn't mine so either you'd have to delegate permission to me to do that, or I give you access to the Docker image and you make the link?

**Reply by flobz** [reply]
> It all seems to run quite well for me
To add change to this repo you need to :
- You need to fork my repo
- then git clone your forked repo
- add your file and commit
- do a git push
- create a pull request on github

**Reply by flobz** [reply]
also to simply the config and facilitate integration in homeassistant we could create a web configuration page.
What do you think ?

**Reply by sOckhamSter** [reply]
Pull request is ready for you...

And yes, I agree a web configuration page would be a good idea. I'm not sure the best way for it to operate so it'll need a bit of thinking. Whether it just runs the required commands sequentially with a web-based CLI presentation, or if there's a fancier way to integrate with the scripts I'm not sure. Something to discuss and have a play around with I guess once we have this version live and working.

**Reply by flobz** [reply]
I merge your PR and did some modification #106.
Thanks !

**Reply by sOckhamSter** [reply]
Thank you. <edited to remove my comment about documentation... I see what you did by moving the file!>

**Reply by sOckhamSter** [reply]
Also, I need to link my Docker image on Docker Hub to your repo. In order to do that I think you need to grant me some form of access to your GitHub repo? I think I need to be added to the organisation/team? Alternatively if you have a Docker account, I can grant you access to the Docker Hub repo and you can delete my 'sOckhamSter' repo link and change it to a 'flobz' one? Is your Docker username flobz?

**Reply by flobz** [reply]
I choose the second option, and I add a github action to automatically build the docker image. Thanks for your contribution.

**Comment by dakoal** [comment]
Hi.

I have a problem running the last docker image on Synology Docker.
everything is running, I see SOC of my e-208 updating and so on, but http requests are running into a 404:

command line output:
2021-05-28 10:41:46,474 :: INFO :: charging status of XXXXXXXXXXXXXXXX is Disconnected, battery level: 83
xxx.xxx.xxx.xxx - - [28/May/2021 10:42:35] "GET / HTTP/1.1" 404 -

On a browser window I get:
Not Found
The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

Started with this:
python3 /psa_car_controller/server.py -f /config/test.json -c /config/charge_config1.json -p 8180 -l 0.0.0.0 -r -b /config

And configured in Docker 8180:8180

**Reply by sOckhamSter** [reply]
I'm also running on Synology. There's no need to change the port that server.py is running on - you can leave -p as 5000 because when it auto-runs after the initial configuration, init.sh will start it on port 5000 anyway and that's not user-configurable. In Synology's Docker interface, just map 8180:5000 instead and then it'll respond on 8180 externally, but use 5000 behind the scenes. Also, you don't need the '-b /config' part.

**Comment by dakoal** [comment]
Great thank you, that was it.
